Lib/fractions.py

```diff
@@ -20,6 +21,39 @@
 # _PyHASH_MODULUS.
 _PyHASH_INF = sys.hash_info.inf
 
+@functools.lru_cache(maxsize = 1 << 14)
+def _hash_algorithm(numerator, denominator):
+
+    # To make sure that the hash of a Fraction agrees with the hash
+    # of a numerically equal integer, float or Decimal instance, we
+    # follow the rules for numeric hashes outlined in the
+    # documentation.  (See library docs, 'Built-in Types').
+
+    try:
+        dinv = pow(denominator, -1, _PyHASH_MODULUS)
```

Comment:
I haven't run a benchmark, yet.  But supposing the inversion here is the bit that's slow, wouldn't it make sense to cache only this part?  That way you get a speedup also if someone merely re-uses denominators.  Not just for the re-use of a complete number?

Reply:
That is slower.  Also, it is a best fit for my known use cases (caching functions that take fractions as arguments, eliminating duplicates with `set()`, and creating singletons).